desc/magnetic_fields.py

```diff
@@ -795,3 +831,464 @@ def odefun(rpz, s):
+    NFP : int
+        number of field periods
+    sym : bool
+        whether to enforce stellarator symmetry. Default is "auto" which enforces if
+        modes are symmetric. If True, non-symmetric modes will be truncated.
+    rho : float [0,1]
+        flux surface label for the toroidal surface
+    name : str
+        name for this surface
+    check_orientation : bool
+        ensure that this surface has a right handed orientation. Do not set to False
+        unless you are sure the parameterization you have given is right handed
+        (ie, e_theta x e_zeta points outward from the surface).
+    surface: FourierRZToroidalSurface, optional, default None
+        Existing FourierRZToroidalSurface object to create a
+        CurrentPotentialField with, if provided will use this
+        object's R_lmn, Z_lmn etc instead of any passed-in values
+
+    """
+
+    _io_attrs_ = (
+        MagneticField._io_attrs_
+        + FourierRZToroidalSurface._io_attrs_
+        + [
+            "_surface_grid",
+            "_params",
+        ]
+    )
+
+    def __init__(
+        self,
+        potential,
+        surface_grid,
+        params={},
+        potential_dtheta=None,
+        potential_dzeta=None,
+        R_lmn=None,
+        Z_lmn=None,
+        modes_R=None,
+        modes_Z=None,
+        NFP=1,
+        sym="auto",
+        rho=1,
+        name="",
+        check_orientation=True,
+        surface=None,
+    ):
+        self._potential = potential
+        self._surface_grid = surface_grid
+        self._params = params
+
+        if surface:
+            if not isinstance(surface, FourierRZToroidalSurface):
+                raise TypeError(
+                    "Expected type FourierRZToroidalSurface for argument surface, "
+                    f"instead got type {type(surface)}"
+                )
+            R_lmn = surface.R_lmn
+            Z_lmn = surface.Z_lmn
+            modes_R = surface._R_basis.modes[:, 1:]
+            modes_Z = surface._Z_basis.modes[:, 1:]
+            NFP = surface.NFP
+            sym = surface.sym
+            rho = surface.rho
+
+        assert surface_grid.NFP == NFP, "NFP of surface must match NFP of surface_grid"
+
+        super().__init__(
+            R_lmn, Z_lmn, modes_R, modes_Z, NFP, sym, rho, name, check_orientation
+        )
+        self._set_derivatives(potential_dtheta, potential_dzeta)
+
+    @property
+    def surface_grid(self):
+        """Grid of points to evaluate surface current at."""
+        return self._surface_grid
+
+    @surface_grid.setter
+    def surface_grid(self, new):
+        if new != self._surface_grid:
+            self._surface_grid = new
+
+    @property
+    def params(self):
+        """Dict of parameters to pass to potential function and its derivatives."""
+        return self._params
+
+    @params.setter
+    def params(self, new):
+        if new != self._params:
+            if len(new) != len(self._params):
+                raise UserWarning(
```

Comment:
you probably want `warnings.warn` here, `raise` will make it an error. 
Could also make a `warnif` util similar to `errorif` to handle the conditional.